server/app/services/applications/ProgramAdminApplicationService.java

```diff
@@ -255,4 +305,71 @@ private Optional<ApplicationModel> validateProgram(
     }
     return application;
   }
+
+  /**
+   * Sets the status on the give applications
+   *
+   * @param applicationList the application list which needs the new status
+   * @param newStatusEvent the StatusEvent carrying the new status
+   * @param admin the admin account initiating the request
+   */
+  public void setStatus(
+      ImmutableList<ApplicationModel> applicationList,
+      StatusEvent newStatusEvent,
+      AccountModel admin)
+      throws StatusNotFoundException, StatusEmailNotFoundException {
+
+    ProgramModel program = applicationList.get(0).getProgram();
+    String newStatusText = newStatusEvent.statusText();
+    // The send/sent phrasing is a little weird as the service layer is converting between intent
+    // and reality.
+    boolean sendEmail = newStatusEvent.emailSent();
+    ProgramDefinition programDef = programRepository.getShallowProgramDefinition(program);
+
+    Optional<Status> statusDefMaybe =
+        applicationStatusesRepository
+            .lookupActiveStatusDefinitions(programDef.adminName())
+            .getStatuses()
+            .stream()
+            .filter(s -> s.statusText().equals(newStatusText))
+            .findFirst();
+    if (statusDefMaybe.isEmpty()) {
+      throw new StatusNotFoundException(newStatusText, program.id);
+    }
+    Status statusDef = statusDefMaybe.get();
+
+    // Send email if requested and present.
+    if (sendEmail) {
+      if (statusDef.localizedEmailBodyText().isEmpty()) {
```

Comment:
Suggestion (non-blocking): Extract sending emails into its own method

Reply:
good idea! I now globalized it for both single application status update and bulk status update :)